### PR TITLE
support TF2

### PR DESCRIPTION
**Pull request by CNOCycle**
solve issue #8 

1. support TF2's keras model

2. detect model's channels ordering and convert ordering automatically

### CODE REVIEW

**Comment by fra31** [comment]
hi,

thanks a lot for the contribution!

I'll try to quickly test and merge it.

File: utils_tf2.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        with tf.GradientTape(watch_accessed_variables=False) as g:
+            g.watch(x_input)
+            logits = self._get_logits(x_input)
+            xent = self._get_xent(logits, y_input)
+        
+        grad_xent = g.gradient(xent, x_input)
+
+        return logits, xent, grad_xent
+
+    def _get_dir(self, logits, y_input):
```
Comment (fra31) [inline utils_tf2.py:67]: `_get_dir` should be renamed `_get_dlr` (and the same below)

**Comment by fra31** [comment]
It looks good to me. Would it be possible for you to add an example (including any small model with checkpoints) of how to use it (as in https://github.com/fra31/auto-attack/tree/master/examples)?

File: utils_tf2.py (hunk 1 above)
Comment (CNOCycle) [reply utils_tf2.py:67]: OK, I will fix format latter.

**Comment by CNOCycle** [comment]
> 
> 
> It looks good to me. Would it be possible for you to add an example (including any small model with checkpoints) of how to use it (as in https://github.com/fra31/auto-attack/tree/master/examples)?

Sure, I will give TF verion's examples.

**Comment by CNOCycle** [comment]
@fra31 I have finished TF version's example

There TF1 version's output:
```
initial accuracy: 97.74%
apgd-ce - 1/10 - 98 out of 1000 successfully perturbed
apgd-ce - 2/10 - 129 out of 1000 successfully perturbed
apgd-ce - 3/10 - 102 out of 1000 successfully perturbed
apgd-ce - 4/10 - 102 out of 1000 successfully perturbed
apgd-ce - 5/10 - 98 out of 1000 successfully perturbed
apgd-ce - 6/10 - 74 out of 1000 successfully perturbed
apgd-ce - 7/10 - 54 out of 1000 successfully perturbed
apgd-ce - 8/10 - 55 out of 1000 successfully perturbed
apgd-ce - 9/10 - 28 out of 1000 successfully perturbed
apgd-ce - 10/10 - 74 out of 774 successfully perturbed
robust accuracy after APGD-CE: 89.60% (total time 47.2 s)
apgd-dlr - 1/9 - 10 out of 1000 successfully perturbed
apgd-dlr - 2/9 - 20 out of 1000 successfully perturbed
apgd-dlr - 3/9 - 15 out of 1000 successfully perturbed
apgd-dlr - 4/9 - 14 out of 1000 successfully perturbed
apgd-dlr - 5/9 - 5 out of 1000 successfully perturbed
apgd-dlr - 6/9 - 5 out of 1000 successfully perturbed
apgd-dlr - 7/9 - 4 out of 1000 successfully perturbed
apgd-dlr - 8/9 - 8 out of 1000 successfully perturbed
apgd-dlr - 9/9 - 10 out of 960 successfully perturbed
robust accuracy after APGD-DLR: 88.69% (total time 105.3 s)
fab - 1/9 - 0 out of 1000 successfully perturbed
fab - 2/9 - 1 out of 1000 successfully perturbed
fab - 3/9 - 2 out of 1000 successfully perturbed
fab - 4/9 - 3 out of 1000 successfully perturbed
fab - 5/9 - 0 out of 1000 successfully perturbed
fab - 6/9 - 3 out of 1000 successfully perturbed
fab - 7/9 - 1 out of 1000 successfully perturbed
fab - 8/9 - 0 out of 1000 successfully perturbed
fab - 9/9 - 2 out of 869 successfully perturbed
robust accuracy after FAB: 88.57% (total time 492.1 s)
square - 1/9 - 31 out of 1000 successfully perturbed
square - 2/9 - 36 out of 1000 successfully perturbed
square - 3/9 - 26 out of 1000 successfully perturbed
square - 4/9 - 34 out of 1000 successfully perturbed
square - 5/9 - 20 out of 1000 successfully perturbed
square - 6/9 - 15 out of 1000 successfully perturbed
square - 7/9 - 20 out of 1000 successfully perturbed
square - 8/9 - 14 out of 1000 successfully perturbed
square - 9/9 - 14 out of 857 successfully perturbed
robust accuracy after SQUARE: 86.47% (total time 727.9 s)
max Linf perturbation: 0.30000, nan in tensor: 0, max: 1.00000, min: 0.00000
robust accuracy: 86.47%
```

There is TF2 version's output
```
initial accuracy: 97.74%
apgd-ce - 1/10 - 101 out of 1000 successfully perturbed
apgd-ce - 2/10 - 127 out of 1000 successfully perturbed
apgd-ce - 3/10 - 106 out of 1000 successfully perturbed
apgd-ce - 4/10 - 106 out of 1000 successfully perturbed
apgd-ce - 5/10 - 96 out of 1000 successfully perturbed
apgd-ce - 6/10 - 73 out of 1000 successfully perturbed
apgd-ce - 7/10 - 53 out of 1000 successfully perturbed
apgd-ce - 8/10 - 55 out of 1000 successfully perturbed
apgd-ce - 9/10 - 31 out of 1000 successfully perturbed
apgd-ce - 10/10 - 76 out of 774 successfully perturbed
robust accuracy after APGD-CE: 89.50% (total time 133.7 s)
apgd-dlr - 1/9 - 8 out of 1000 successfully perturbed
apgd-dlr - 2/9 - 21 out of 1000 successfully perturbed
apgd-dlr - 3/9 - 6 out of 1000 successfully perturbed
apgd-dlr - 4/9 - 13 out of 1000 successfully perturbed
apgd-dlr - 5/9 - 4 out of 1000 successfully perturbed
apgd-dlr - 6/9 - 9 out of 1000 successfully perturbed
apgd-dlr - 7/9 - 6 out of 1000 successfully perturbed
apgd-dlr - 8/9 - 4 out of 1000 successfully perturbed
apgd-dlr - 9/9 - 7 out of 950 successfully perturbed
robust accuracy after APGD-DLR: 88.72% (total time 297.1 s)
fab - 1/9 - 0 out of 1000 successfully perturbed
fab - 2/9 - 2 out of 1000 successfully perturbed
fab - 3/9 - 1 out of 1000 successfully perturbed
fab - 4/9 - 3 out of 1000 successfully perturbed
fab - 5/9 - 0 out of 1000 successfully perturbed
fab - 6/9 - 3 out of 1000 successfully perturbed
fab - 7/9 - 1 out of 1000 successfully perturbed
fab - 8/9 - 0 out of 1000 successfully perturbed
fab - 9/9 - 2 out of 872 successfully perturbed
robust accuracy after FAB: 88.60% (total time 1808.1 s)
square - 1/9 - 28 out of 1000 successfully perturbed
square - 2/9 - 35 out of 1000 successfully perturbed
square - 3/9 - 33 out of 1000 successfully perturbed
square - 4/9 - 33 out of 1000 successfully perturbed
square - 5/9 - 21 out of 1000 successfully perturbed
square - 6/9 - 12 out of 1000 successfully perturbed
square - 7/9 - 19 out of 1000 successfully perturbed
square - 8/9 - 13 out of 1000 successfully perturbed
square - 9/9 - 18 out of 860 successfully perturbed
robust accuracy after SQUARE: 86.48% (total time 2051.4 s)
max Linf perturbation: 0.30000, nan in tensor: 0, max: 1.00000, min: 0.00000
robust accuracy: 86.48%
```

File: examples/eval_tf2.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+import numpy as np
+import tensorflow as tf
+
+import torch
+import torch.nn as nn
+import torchvision.datasets as datasets
+import torch.utils.data as data
+import torchvision.transforms as transforms
+
```
Comment (fra31) [inline examples/eval_tf2.py:12]: it needs
``` python
import sys
sys.path.insert(0, '..')
```

File: utils_tf2.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        dlr_target = dlr_loss_targeted(logits, y_input, y_target, num_classes=self.num_classes)
+        return dlr_target
+
+    def __get_grad_dir_target(self, x_input, y_input, y_target):
+        with tf.GradientTape(watch_accessed_variables=False) as g:
+            g.watch(x_input)
+            logits = self.__get_logits(x_input)
+            dlr_target = self.__get_dir_target(logits, y_input, y_target)
+
+        grad_target = g.gradients(dlr_target, x_input)[0]
```
Comment (fra31) [inline utils_tf2.py:91]: `grad_target = g.gradient(dlr_target, x_input)`

File: examples/eval_tf1.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+import numpy as np
+import tensorflow as tf
+
+import torch
+import torch.nn as nn
+import torchvision.datasets as datasets
+import torch.utils.data as data
+import torchvision.transforms as transforms
+
```
Comment (fra31) [inline examples/eval_tf1.py:12]:
```python
import sys
sys.path.insert(0, '..')
```

**Comment by fra31** [comment]
I finally ran the examples and everything works fine. I just noted a few small things above, and, once fixed them, I think it's ready to be merged. Do you have any idea why with TF2 is almost 3x slower than with TF1?

**Comment by CNOCycle** [comment]
I fixed those bugs.

I also noticed that execution time of TF2 is much slower than that of TF1.

However, execution time is no difference when evaluating "initial accuracy".

I cannot give you a reason explain why the performance has significant gap because pytorch and TF version are totally different in my testing environment.

In my opinion, I thought attacks work properly and this PR can be merged first.

**Comment by CNOCycle** [comment]
I thought pytorch side may be improved. Because the heaviest computations have done in TF side on GPU, those data may not need to transfer to GPU again in pytorch side.

**Comment by fra31** [comment]
Thanks for the quick updates and for the contribution!

I will investigate further the difference between the runtime in TF1 and TF2.
